src/Symfony/Component/HttpFoundation/InputBag.php

```diff
@@ -0,0 +1,96 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpFoundation;
+
+use Symfony\Component\HttpFoundation\Exception\BadRequestException;
+
+/**
+ * InputBag is a container for user input values such as $_GET, $_POST, $_REQUEST, and $_COOKIE.
+ *
+ * @author Saif Eddin Gmati <saif.gmati@symfony.com>
+ */
+final class InputBag extends ParameterBag
+{
+    /**
+     * Returns a string input value by name.
+     *
+     * @param string|null $default The default value if the input key does not exist
+     *
+     * @return string|null
+     */
+    public function get(string $key, $default = null)
+    {
+        if (null !== $default && !is_scalar($default) && !\is_callable([$default, '__toString'])) {
+            @trigger_error(sprintf('Passing a non-string value as 2nd argument to "%s()" is deprecated since Symfony 5.1, pass a string or null instead.', __METHOD__), E_USER_DEPRECATED);
+        }
+
+        $value = parent::get($key, $this);
+
+        if (null !== $value && $this !== $value && !is_scalar($value) && !\is_callable([$value, '__toString'])) {
+            @trigger_error(sprintf('Retrieving a non-string value from "%s()" is deprecated since Symfony 5.1, and will throw a "%s" exception in Symfony 6.0, use "%s::all()" instead.', __METHOD__, BadRequestException::class, __CLASS__), E_USER_DEPRECATED);
+        }
+
+        return $this === $value ? $default : $value;
+    }
+
+    /**
+     * Returns the inputs.
+     *
+     * @param string|null $key The name of the input to return or null to get them all
+     */
+    public function all(string $key = null): array
+    {
+        if (null === $key) {
+            return $this->parameters;
+        }
+
+        $value = $this->parameters[$key] ?? [];
+        if (!\is_array($value)) {
+            throw new Exception\BadRequestException(sprintf('Unexpected value for "%s" input, expecting "array", got "%s".', $key, \is_object($value) ? \get_class($value) : \gettype($value)));
```

Comment:
If that returns a 400 error code using this exception, then https://github.com/symfony/symfony/blob/2f1946feab3a3d5cd38152ce8576a8c4b865da92/src/Symfony/Component/HttpFoundation/HeaderBag.php#L201-L203 should do for consistency.
And https://github.com/symfony/symfony/blob/2f1946feab3a3d5cd38152ce8576a8c4b865da92/src/Symfony/Component/HttpFoundation/ParameterBag.php#L153-L156 does not throw any exception at all if people pass a non-parsable bool value.
In our apps we had to built a helper class to return 400 error in those cases. While I agree that it should ideally throw an exception resulting in 400 http error, the reality is that this is not how all the parameter bags and header bags etc behave so far. The header bag can also contain user input. So the current solution is not consistent and half-done to me.

Reply:
Can you be more precise with what you mean with "half-done" here?
Are the examples you gave the full "other half"? If not what's missing?
Anyway, this looks loosely related to me - it could be done in a separate PR.
This one is a big enough selfconsistent change IMHO.

Reply:
@Tobion i agree that other getters ( e.g `getInt` ) should be fixed to throw a bad request exception too. however, this won't be done in this PR ( and i am not sure if we should deprecate getting non-`x` values from `getX` first )

Reply:
> Can you be more precise with what you mean with "half-done" here?

When we say that the parameter bags are meant to throw a 400 if invalid parameters are passed like this PR starts, then all the getters should behave like this like getInt, getBoolean etc. Otherwise it's not consistent to me. If you think it's ok to tackle this later, that's fine for me. 
> What's in the helper you mention?
Alternatives for getInt, getBoolean, getArray etc to throw a 400.

Reply:
>When we say that the parameter bags are meant to throw a 400 if invalid parameters are passed like this PR starts, then all the getters should behave like this like getInt, getBoolean etc. Otherwise it's not consistent to me. If you think it's ok to tackle this later, that's fine for me.

> Alternatives for getInt, getBoolean, getArray etc to throw a 400.

currently the trigger a deprecation because of the use of `get` and will result in exception in 6.0, so what is the issue here ?